research/gaia/gremlin/gs_gremlin/src/graph_proxy/global_storage.rs

```diff
@@ -0,0 +1,454 @@
+//! limitations under the License.
+
+use crate::graph_proxy::util::{str_to_dyn_error, IterList};
+use dyn_type::{Object, Primitives};
+use gremlin_core::graph_proxy::from_fn;
+use gremlin_core::structure::LabelId as RuntimeLabelId;
+use gremlin_core::structure::{
+    DefaultDetails, Direction, DynDetails, Edge, Label, PropKey, QueryParams, Statement, Vertex,
+};
+use gremlin_core::{filter_limit, filter_limit_ok, limit_n};
+use gremlin_core::{register_graph, DynResult, GraphProxy, ID};
+use maxgraph_store::api::graph_partition::GraphPartitionManager;
+use maxgraph_store::api::graph_schema::Schema;
+use maxgraph_store::api::prelude::Property;
+use maxgraph_store::api::PropId;
+use maxgraph_store::api::*;
+use maxgraph_store::api::{Edge as StoreEdge, Vertex as StoreVertex};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+static INVALID_LABEL_ID: LabelId = 0xffffffff;
+static INVALID_PROP_ID: PropId = 0xffffffff;
+// TODO(bingqing): confirm with compiler, compiler should set the param_key as "SID"
+const SNAPSHOT_ID: &str = "SID";
+
+pub struct GraphScopeStore<V, VI, E, EI>
+where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    store: Arc<dyn GlobalGraphQuery<V = V, E = E, VI = VI, EI = EI>>,
+    partition_manager: Arc<dyn GraphPartitionManager>,
+}
+
+#[allow(dead_code)]
+pub fn create_gs_store<V, VI, E, EI>(
+    store: Arc<dyn GlobalGraphQuery<V = V, E = E, VI = VI, EI = EI>>,
+    partition_manager: Arc<dyn GraphPartitionManager>,
+) where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    let graph = GraphScopeStore {
+        store,
+        partition_manager,
+    };
+    register_graph(Arc::new(graph));
+}
+
+impl<V, VI, E, EI> GraphProxy for GraphScopeStore<V, VI, E, EI>
+where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    fn scan_vertex(
+        &self,
+        params: &QueryParams<Vertex>,
+    ) -> DynResult<Box<dyn Iterator<Item = Vertex> + Send>> {
+        let store = self.store.clone();
+        let si = params
+            .get_extra_param(SNAPSHOT_ID)
+            .ok_or(str_to_dyn_error("get snapshot_id failed"))?
+            .as_i64()
+            .map_err(|e| str_to_dyn_error(&e.to_string()))? as SnapshotId;
+        let schema = store
+            .get_schema(si)
+            .ok_or(str_to_dyn_error("get schema failed"))?;
+        let label_ids = encode_storage_label(params.labels.as_ref(), schema.clone());
+        let prop_ids = encode_storage_prop_key(params.props.as_ref(), schema.clone());
+        let filter = params.filter.clone();
+        let result = store
+            .get_all_vertices(
+                si,
+                label_ids.as_ref(),
+                None,
```

Comment:
Can common these two `None` to increase readability.

Reply:
Done